docs: platform portability and infrastructure documentation (#91237)
vercel/next.js: docs/01-app/02-guides/how-revalidation-works.mdx

---
title: How revalidation works in Next.js
nav_title: How Revalidation Works
description: A deep dive into how Next.js revalidates cached content, including the tag system, cache consistency, and multi-instance coordination.
related:
  description: Related guides and references.
  links:
    - app/getting-started/caching
    - app/guides/incremental-static-regeneration
    - app/api-reference/config/next-config-js/cacheHandlers
    - app/api-reference/functions/revalidateTag
---

> The [Caching](/docs/app/getting-started/caching) page covers how to use `use cache`, `cacheTag`, and `cacheLife`. This page explains **how revalidation works internally**, for platform engineers and advanced users who need to understand the system to implement [custom cache handlers](/docs/app/api-reference/config/next-config-js/cacheHandlers) or debug revalidation behavior.

Reviewer: Should this also cover "older APIs":
 - Page router's on-demand revalidation (AFAIK can't be satisfied with cacheHandlers and you need to either set up things to work with `x-prerender-revalidate` setup or provide custom `revalidate()` implementation just for it)
 - Patched fetch (+ `unstable_cache`, which still exist) seem to need the older `CacheHandler` instance/interface
?

Reviewer 2: Updated: added note covering Pages Router on-demand ISR APIs (res.revalidate and x-prerender-revalidate) and clarified they use cacheHandler (singular).

## The Revalidation Model

> Every server-rendered route in Next.js can be revalidated on demand. The ability to update any cached content without redeploying is a core part of Next.js's [rendering model](/docs/app/guides/rendering-philosophy).

Reviewer 3: Every route, right? I'm not sure what server-rendered is supposed to mean here, since AFAIK quite literally every route can be revalidated on demand.

```suggestion
Every route in Next.js can be revalidated on demand. The ability to update any cached content without redeploying is a core part of Next.js's [rendering model](/docs/app/guides/rendering-philosophy).
```

Reviewer 4: Yeah indeed

Reviewer 2: You can't revalidate automatically statically optimized pages (from pages router) since they aren't a "prerender" and are considered purely static.

Reviewer 2: Updated to "Most routes" and explicitly documented Pages Router auto-static pages as not on-demand revalidated.

There are two types of revalidation:

- **Time-based revalidation** uses a stale-while-revalidate pattern. The cached content is served immediately, and a background regeneration is triggered when the content's age exceeds the [`cacheLife`](/docs/app/api-reference/functions/cacheLife) or `revalidate` duration. The stale content continues to be served until the fresh content is ready.
- **On-demand revalidation** explicitly invalidates cached content by calling [`revalidateTag()`](/docs/app/api-reference/functions/revalidateTag) or [`revalidatePath()`](/docs/app/api-reference/functions/revalidatePath). The next request to that content triggers a fresh render.

## What Gets Revalidated

When a route is revalidated, Next.js regenerates **both** the HTML response and the RSC payload (React Server Components payload) from the same React component tree. Both artifacts are stored together in the same cache entry.

> This consistency matters because the RSC payload carries the component tree structure, not just data.

Reviewer 4: ```suggestion
This consistency matters because the RSC payload is used for client-side navigations. Browser navigations and client-side navigations should hold the same content.
```

### What happens if they get out of sync

> If a platform's cache serves HTML from one render and an RSC payload from a different render, users may see stale or mismatched content during client-side navigation. The primary mitigation is to cache HTML and RSC responses together with the same TTL and invalidation policy, and to respect the [`Vary` header](/docs/app/guides/cdn-caching) that Next.js sets. See [CDN Caching](/docs/app/guides/cdn-caching) for details.

Reviewer 3: ```suggestion
If a platform's cache serves HTML from one render and an RSC payload from a different render, users may see stale or mismatched content during client-side navigation. The primary mitigation is to cache HTML and RSC responses together with the same TTL and invalidation policy, and to respect the [`Vary` header](/docs/app/guides/cdn-caching) that Next.js sets. Note that is not compatible with [standard HTTP caching](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/Caching); see [CDN Caching](/docs/app/guides/cdn-caching) for details.
```

Reviewer 2: We're linking to that mozilla doc saying Vary isn't standard caching but that doc does mention vary in there for CDN cache.

Reviewer 2: Kept the CDN Caching guidance link and did not add the "not compatible with standard HTTP caching" sentence, aligned with the follow-up discussion.

A separate but related problem is **cross-deployment skew**: during rolling deployments, a client built with deploy A may receive responses from a server running deploy B. [`deploymentId`](/docs/app/api-reference/config/next-config-js/deploymentId) mitigates this: when the client detects a different deployment ID from the server, it triggers a hard navigation to fetch consistent content.

## Tag System Architecture

> Next.js uses a tag-based system to track which cached content needs to be invalidated. There are two types of tags:

Reviewer 4: ```suggestion
Next.js uses a tag-based system to track which cached content needs to be invalidated. There are two types of tags:

- Explicit Tags
- Soft Tags
```

Reviewer 2: Updated: section now explicitly states the two tag types (explicit and soft).

### Explicit tags

Explicit tags are set by the developer using [`cacheTag()`](/docs/app/api-reference/functions/cacheTag) inside a `use cache` function, or via `next: { tags: [...] }` on a `fetch` call. When [`revalidateTag('my-tag')`](/docs/app/api-reference/functions/revalidateTag) is called, all cache entries with that tag are invalidated.

### Soft tags

Soft tags are automatically generated by Next.js based on the route path, prefixed with `_N_T_`. For example, the route `/blog/hello` generates soft tags like `_N_T_/layout`, `_N_T_/blog/layout`, `_N_T_/blog/hello/layout`, and `_N_T_/blog/hello`. Each segment in the path gets a layout tag, plus the leaf route itself.

> Soft tags enable [`revalidatePath()`](/docs/app/api-reference/functions/revalidatePath) to work through the same tag-based system. When `revalidatePath('/blog/hello')` is called, it invalidates all cache entries associated with that path's soft tags.

Reviewer 4: > it invalidates all cache entries associated with that path's soft tags.

It's left ambiguous if that includes the `_N_T_/blog/layout` etc.

Worth clarifying here.

Reviewer 2: Updated: clarified that revalidatePath includes ancestor layout soft tags and added concrete examples.
